File: src/main/java/xyz/nkomarn/composter/network/protocol/packet/serverbound/ServerboundHandshakePacket.java
package xyz.nkomarn.composter.network.protocol.packet.serverbound;

import xyz.nkomarn.composter.network.WrappedBuff;
import xyz.nkomarn.composter.network.protocol.handler.PacketHandler;

public class ServerboundHandshakePacket implements ServerboundPacket {

    private String username;

    public String username() {
        return username;
    }

    @Override
    public void decode(WrappedBuff buff) {
        username = buff.readString();
    }

    @Override
    public void handle(PacketHandler handler) {
        handler.handleHandshake(this);
    }
}

File: src/main/kotlin/kyta/composter/protocol/packet/handshaking/ServerboundHandshakePacket.kt
package kyta.composter.protocol.packet.handshaking

import kyta.composter.protocol.PacketHandler
import kyta.composter.protocol.PacketSerializer
import kyta.composter.protocol.ReadBuffer
import kyta.composter.protocol.ServerboundPacket

data class ServerboundHandshakePacket(val username: String) : ServerboundPacket {
    override suspend fun handle(handler: PacketHandler) = handler.handleHandshake(this)

    companion object : PacketSerializer<ServerboundHandshakePacket> {
        override fun deserialize(buffer: ReadBuffer): ServerboundHandshakePacket {
            return ServerboundHandshakePacket(buffer.readString())
        }
    }
}